thriftbp/server.go

```diff
@@ -91,38 +96,232 @@ func NewServer(cfg ServerConfig) (*thrift.TSimpleServer, error) {
+
+// buildConfig creates a new ServerConfig instance and applies
+// all the supplied configuration options to it.
+func buildConfig(options ...ServerOption) (ServerConfig, error) {
+	cfg := &ServerConfig{}
+	for _, option := range options {
+		err := option(cfg)
+		if err != nil {
+			return ServerConfig{}, err
+		}
+	}
+
+	return *cfg, nil
+}
+
+// withConfigFrom returns a server option that overwrites the
+// any existing values in a ServerConfig with those defined
+// in src. This is meant to be used from NewBaseplateServer
+// for mapping the supplied ServerConfig into a ServerOption
+func withConfigFrom(src ServerConfig) ServerOption {
+	return func(dst *ServerConfig) error {
+		*dst = src
+		return nil
+	}
+}
+
+// DefaultServerOptions builds and returns a slice of the default
+// baseplate server config options. This method is meant to support
+// advanced use cases when building a custom baseplate server.
+func DefaultServerOptions(bp baseplate.Baseplate) []ServerOption {
+	return []ServerOption{
+		withDefaultMiddleware(bp),
+		withDefaultLogger(bp),
+		withDefaultListenerAddress(bp),
+		withDefaultSocket(),
+	}
+}
+
+// withDefaultMiddleware will prepend the default middleware so
+// we can maintain backwards compatibility with the old logic of
+// NewBaseplateServer.
+func withDefaultMiddleware(bp baseplate.Baseplate) ServerOption {
+	return func(cfg *ServerConfig) error {
+		existingMiddleware := cfg.Middlewares
+
+		if err := WithDefaultMiddleware(bp)(cfg); err != nil {
+			return err
+		}
+
+		// Prepend defaults to existing values
+		cfg.Middlewares = append(cfg.Middlewares, existingMiddleware...)
+		return nil
+	}
+}
+
+// WithDefaultMiddleware will append the default baseplate middleware to the
+// server config. When depending on specific features such as ErrorSpanSuppressor,
+// the ServerOption configuring the setting needs to be invoked before this method.
+func WithDefaultMiddleware(bp baseplate.Baseplate) ServerOption {
+	return func(cfg *ServerConfig) error {
+		return WithMiddleware(BaseplateDefaultProcessorMiddlewares(
+			DefaultProcessorMiddlewaresArgs{
+				EdgeContextImpl:                    bp.EdgeContextImpl(),
+				ErrorSpanSuppressor:                cfg.ErrorSpanSuppressor,
+				ReportPayloadSizeMetricsSampleRate: cfg.ReportPayloadSizeMetricsSampleRate,
+			},
+		)...)(cfg)
+	}
+}
+
+func withDefaultLogger(bp baseplate.Baseplate) ServerOption {
+	return WithDefaultLogger(bp.GetConfig().Log.Level)
+}
+
+// withDefaultListenerAddress is used to copy the address config from the
+// baseplate config to the server config. Currently, it's only
+// needed to guarantee backwards compatibility in NewServer
+func withDefaultListenerAddress(bp baseplate.Baseplate) ServerOption {
+	return WithListenerAddress(bp.GetConfig().Addr)
+}
+
+// WithMiddleware appends the arg middleware to the end of
+// the cfg.Middleware slice
+func WithMiddleware(middleware ...thrift.ProcessorMiddleware) ServerOption {
```

Comment:
I can see this API being confusing and becomes a footgun. All the other `With*` APIs are to replace the corresponding config field, while this one is to append.